http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -140,20 +139,30 @@ class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Concurrent](
               savedMasterMachineType match {
                 case Some(machineType) =>
                   for {
+                    runtimeConfig <- dbRef.inTransaction(
+                      RuntimeConfigQueries.getRuntimeConfig(resolvedCluster.runtimeConfigId)
+                    )
                     // perform gddao and db updates for new resources
-                    _ <- Async[F].liftIO(clusterHelper.updateMasterMachineType(resolvedCluster, machineType))
-                    now <- Timer[F].clock.realTime(TimeUnit.MILLISECONDS)
+                    _ <- runtimeConfig.cloudService match {
```

Comment:
I wonder if it would make sense to have a type class something like 
```
trait RuntimeAlg[F] {
   def updateMachineType(UpdateMachineTypeParams(resolvedCluster, machineType, now)): F[Unit]
   def startRuntime(StartRuntimeParams(resolvedCluster, now)): F[Unit]
/// other common functions
}
```

And then we can do ```runtimeAlg.updateMachineType(xxx)``` etc

Reply:
Trying to understand how a typeclass would work here. With a typeclass I'd summon the instances like this:
```
cloudService match {
  case GCE => RuntimeAlg[F, GCE].updateMachineType(..)
  case Dataproc => RuntimeAlg[F, Dataproc].updateMachineType(..)
}
```
That might be a good way to do it instead of passing each interpreter to the constructor. But I don't see how it replaces the abstract class `BaseRuntimeInterpter` which is used for common functionality between GCE and Dataproc. Am I missing something?

Reply:
maybe not get rid of `BaseRuntimeInterpter `, this is what I was thinking roughly

```
trait RuntimeAlgebra[F[_]] {
  def createRuntime(params: CreateRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[CreateRuntimeResponse]
  def deleteRuntime(params: DeleteRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def finalizeDelete(params: FinalizeDeleteParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def stopRuntime(params: StopRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def startRuntime(params: StartRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def updateMachineType(params: UpdateMachineTypeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def updateDiskSize(params: UpdateDiskSizeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
}

// instances
object RuntimeInstances {
 def gce[F[_]: Async: ContextShift: Logger: DbReference: NewRelicMetrics](config: RuntimeInterpreterConfig.GceInterpreterConfig,
                                            welderDao: WelderDAO[F])(implicit ec: ExecutionContext): RuntimeAlgebra[F] = new BaseRuntimeInterpreter(config, welderDao){
   override protected def stopGoogleRuntime(runtime: Runtime, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override protected def startGoogleRuntime(runtime: Runtime, welderAction: WelderAction, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override protected def setMachineTypeInGoogle(runtime: Runtime, machineType: MachineTypeName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override def createRuntime(params: CreateRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[CreateRuntimeResponse] = ???

   override def deleteRuntime(params: DeleteRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override def finalizeDelete(params: FinalizeDeleteParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override def updateDiskSize(params: UpdateDiskSizeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???
 }
  
  def dataproc[F[_]: Async: ContextShift: Logger: DbReference: NewRelicMetrics](config: RuntimeInterpreterConfig.DataprocInterpreterConfig,
                                                                                welderDao: WelderDAO[F])(implicit ec: ExecutionContext): RuntimeAlgebra[F] = new BaseRuntimeInterpreter(config, welderDao){
    override protected def stopGoogleRuntime(runtime: Runtime, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override protected def startGoogleRuntime(runtime: Runtime, welderAction: WelderAction, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override protected def setMachineTypeInGoogle(runtime: Runtime, machineType: MachineTypeName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override def createRuntime(params: CreateRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[CreateRuntimeResponse] = ???

    override def deleteRuntime(params: DeleteRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override def finalizeDelete(params: FinalizeDeleteParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override def updateDiskSize(params: UpdateDiskSizeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???
  }
}
```

Reply:
I think I see.. want to see how the code looks